lib/recog/nizer.rb

```diff
@@ -20,43 +20,122 @@ class Nizer
   #
   # @param match_key [String] Fingerprint DB name, e.g. 'smb.native_os'
-  # @return (see Fingerprint#match)
+  # @param match_string [String] String to match
+  # @return (see Fingerprint#match) or nil
   def self.match(match_key, match_string)
-    match_string = match_string.to_s.unpack("C*").pack("C*")
-    @@db_manager ||= Recog::DBManager.new
-    @@db_manager.databases.each do |db|
-      next unless db.match_key == match_key
-      db.fingerprints.each do |fprint|
-        m = fprint.match(match_string)
-        return m if m
-      end
-    end
+    filter = { match_key: match_key, multi_match: false }
+    matches = self.match_all_db(match_string, filter)
+
+    return matches[0] if matches && matches[0]
+
     nil
   end
 
+  #
+  # @param match_key [String] Fingerprint DB name, e.g. 'smb.native_os'
+  # @param match_string [String] String to match
+  # @return [Array] Array of Fingerprint#match or empty array
   def self.multi_match(match_key, match_string)
-    match_string = match_string.to_s.unpack("C*").pack("C*")
-    @@db_manager ||= Recog::DBManager.new
+    filter = { match_key: match_key, multi_match: true }
+    matches = self.match_all_db(match_string, filter)
 
-    matches = Array.new #array to hold all fingerprint matches
+    matches
+  end
 
-    @@db_manager.databases.each do |db|
-      next unless db.match_key == match_key
+  #
+  # Search all fingerprint dbs and attempt to find matching
+  # {Fingerprint fingerprint}s. It will return the first match found
+  # unless the :multi_match option is used to request all matches.
+  # Returns an array of all matching fingerprints or an empty array.
+  #
+  # @param match_string [String] Service banner to match
+  # @param [Hash] filters This hash contains filters used to limit the
+  #   results to just those from specific types of fingerprints.
+  #   The values that these filters match come from the 'fingerprints' top
+  #   level element in the fingerprint DB XML or, in the case of 'protocol',
+  #   this value can be overridden at the individual fingerprint level by
+  #   setting a value for 'service.protocol'
+  #
+  #   With the exception of 'match_key', the filters below match the
+  #   'fingerprints' attributes with the same name.
+  # @option filters [String] :match_key Value from XML 'matches' or file name
+  # @option filters [String] :database_type fprint db type: service, util.os, etc.
+  # @option filters [String] :protocol Protocol (ftp, smtp, etc.)
+  # @option filters [String] :multi_match Return all matches instead of first
+  # @return [Array] Array of Fingerprint#match or empty array
+  def self.match_all_db(match_string, filters = {})
+    match_string = match_string.to_s.unpack('C*').pack('C*')
+    matches = Array.new # array to hold all fingerprint matches
 
+    self.load_db unless @@db_manager
+
+    return matches unless !@@db_manager.databases || !@@db_manager.databases.empty?
```

Comment:
I think the first part of this `unless` represents a bug, but if nothing else this is confusing to read.  I think it would be cleaner to ensure that db_manager initializes `databases` to an empty array.  Then you can remove this line entirely. `databases` will never be `nil`, and it is fine to iterate over it even if empty since you still just return the empty `matches`.